### ISSUE TITLE
Fix zero-lamport accounts preventing slot cleanup

### ISSUE THREAD

**Issue by jeffweshington** [solana-labs-community/solana#118]
#### Problem
Some slot `S` containing an update for an account with key `K` may not be cleaned up if at the time of `clean_slots`, the latest update in `K` is in some rooted slot `S' > S` and `K` has zero lamports. This is because in `clean_slots()`

`if account_info.lamports == 0 {`

and

`else if accounts_index.uncleaned_roots.contains(slot)`

are mutually exclusive, and `uncleaned_roots` is purged at the end of the clean_slots() loop, so those slots will not be re-examined if the pubkey is not updated again in some future root that is added to `uncleaned root`

#### Summary of Changes
Fixes #

**Comment by t-neIson** [comment]
runtime/src/accounts_db.rs
```rust
                        let (slot, account_info) = &list[index];
                        if account_info.lamports == 0 {
                            purges.insert(*pubkey, accounts_index.would_purge(pubkey));
                        } else if accounts_index.uncleaned_roots.contains(slot) {
```
Thanks for fixing another AccountsDB bug.. ;)

iirc, there was some reason that I intentionally chained with `else if`. But I can't remember the exact reason. And, sorry for the lack of comment. I'll try to remember and report back here. Anyway, a lot happened since then. So, this is just ok as long as the ci is green and running validator causes no bad snapshot.

**Comment by sakeridge** [comment]
no worries! it seems some of other tests are failing. Could you look at it? Those failing tests will be very cryptic to you. So don't hesitate to ask questions. At first glance, I can't tell these are actually failing by being broken or these test assertions need to just be updated.

**Comment by mvilnes** [comment]
As said in the my above comment, I don't think my thought for the original implementation decision started from these concerns. At the time, I didn't even tried to do `if` approach. Anyway, I think your observation is correct for the interrelation of two cleaning and panic of wrap with the `if` approach..

**Comment by CrieasofCarrots** [comment]
Yeah, working on those tests! I think the ones that are failing are a mix of assertions failing and me not handling: #12606 (comment), which is breaking the issue described here: #7157 (This is the issue with old AppendVecs reviving zero-lamport accounts, which should now be addressed by `calc_delete_dependencies()`

**Comment by jeffweshington** [comment]
I see! It looks that you passed the CI fully. So happy. :)

So, the once independent cleaning mechanisms finally start to interact each other in one pass after battle-tested for half a year. That sounds cool.
